PR: 最小限構成でのメモアプリ実装

kei615ykhm opened the pull request:
## 概要
このプルリクエストでは、`feature/minimal-configuration`ブランチから`develop`ブランチへの統合を行います。主な目的は、簡易的な最小限構成でのメモアプリを作成し、今後のアップデートの土台を構築することです。

＊こちらは、コードレビューのご依頼とマージ認証を目的としたPRになります。

## 作業内容
- 型定義の作成 (`src/types/index.ts`)
- カスタムフックの作成 (`src/hooks/useMemos.ts`)
- メモフォームコンポーネントの作成 (`src/components/memos/MemoForm.tsx`)
- メモリストコンポーネントの作成 (`src/components/memos/MemoList.tsx`)
- メインページの作成 (`src/app/page.tsx`)

## 実装の変更点
issue #9 にて当初予定していたローカルストレージを使用した保存機能の実装は、以下の理由により保留としました：

1. ローカルストレージの実装と`useEffect`の使用に関して、再学習が必要だと判断したため
2. 最終的な目標が`Supabase`のデータベースを使用した保存機能であるため、学習効率を考慮し、ローカルストレージでの実装に時間をかけるべきではないと判断したため

詳細: Issue #10 （クローズ済み）
作業ブランチ: `experimental/use-local-server`

## 今後の予定
1. ESLintの設定と必要に応じたコード修正（Issue #11）
2. ダッシュボード機能の実装（Issue #12 ）
   - Parallel Routesを使用したダッシュボードレイアウトの作成
   - メモリスト機能の実装（検索機能含む）
   - メモ作成機能の実装
3. SWR/SSG対応の下地作り（Issue #13 ）
   - モーダルコンポーネントの作成（将来の実装用）
   - SWRの下地作り
4. Supabaseとの連携によるデータベース実装
5. 保存機能の実装
6. メモの編集機能の追加
7. メモの詳細ページの作成

なお、具体的な実装詳細については別途Issueを立ち上げ、そちらで管理します。

## レビュー時の注意点
- 現時点では編集と保存機能は実装されていないため、アプリの再読み込み時にデータが失われます
- コンポーネントの構造と型定義が適切かどうかご確認ください
- カスタムフックの実装が今後の拡張性を考慮したものになっているかご確認ください

## テスト
- [ ] メモの追加が正常に動作すること
- [ ] メモリストが正しく表示されること
- [ ] コンポーネント間のデータの受け渡しが適切に行われていること

ご確認よろしくお願いいたします。

kei615ykhm committed:
remove: ファビコンファイルの削除

kei615ykhm committed:
add: #9 のタスクである型定義の作成

kei615ykhm committed:
add: #9 のタスクであるカスタムフックの作成

- ここからメモの状態管理とCRUD操作を行うためのコードを書き始める

kei615ykhm committed:
add: #9 メモの状態を追加

kei615ykhm committed:
add: #9 `useState`と`types/index.ts`に定義したMemo型をそれぞれインポート

kei615ykhm committed:
add: #9 `addMemo`関数を追加してメモを追加する機能を作成

kei615ykhm committed:
add: #9 `deleteMemo`関数を追加してメモを削除する機能を作成

kei615ykhm committed:
add: #9 `saveMemo`関数を追加してローカルストレージに保存する機能を作成

kei615ykhm committed:
add: `useEffect`の基本構造を記述

kei615ykhm committed:
add: #9 `useEffect`を追加してローカルストレージの初期ロードを作成

kei615ykhm committed:
remove: #9 作業のために初期状態だったコードをすべて削除

- これから動作確認のためのページを作成する

kei615ykhm committed:
add: #9 Homeコンポーネント内にて`CSR`設定と`useMemos`フックを呼び出す

- コードの途中であるため描画なし

kei615ykhm committed:
add: #9 基本的な`UI`構造の作成

kei615ykhm committed:
add: #9 動作確認のためにテストメモを追加する機能を作成

- `onClick`を設定し、ボタンがクリックされたときに`addMemo`関数が呼び出される

kei615ykhm committed:
add: #9 動作確認のためにテストメモを削除する機能を作成

- `onClick`を設定し、ボタンがクリックされたときに`deleteMemo`関数が呼び出される

kei615ykhm committed:
add: フォームとリストを動作確認しながら作成するための`import`とTODOを記述

kei615ykhm committed:
add: メモフォームコンポーネントのファイルとひな形を作成

kei615ykhm committed:
add: #9 フォームの基本構造と送信処理のTODOを記録

kei615ykhm committed:
add: typeScriptの型定義

- `useState`で、contentという状態変数とその値を更新するために`setContent`関数を定義

kei615ykhm committed:
add: `content`が保持する値を自動更新する記述

- ユーザーが入力した値に応じて自動的に入力フィールドの表示を更新する

kei615ykhm committed:
add: 入力値を`content`に反映させる記述

- ユーザーの入力をリアルタムで補足してReactの状態として保持する

kei615ykhm committed:
docs: データフローをコメントで補助的に説明

kei615ykhm committed:
add: `props`の`interface`を定義

kei615ykhm committed:
update: `onSubmit`の誤植を修正

kei615ykhm committed:
docs: データフローをコメントで補助的に説明

kei615ykhm committed:
add: #9 フォーム送信時の処理を実行する関数を定義

kei615ykhm committed:
add: #9 フォーム送信時の処理を実装

- ページリロードなどの防止し、有効な入力があったときの挙動指示し、入力後にフィールドをクリアにする

kei615ykhm committed:
update: #9 フォーム送信時の処理を改良

- 安全性の観点から、空白のみの入力をチェックして中身があれば処理を続行するように設定を追加した

kei615ykhm committed:
update: #9 フォーム送信時の処理関数を定義

kei615ykhm committed:
move: `MemoForm`を`components/memos`配下に移動

kei615ykhm committed:
remove: デフォルトインポートを削除

kei615ykhm committed:
update: インポートの指定パスを変更

- `MemoForm`が`components/memos`配下に移動したため

kei615ykhm committed:
remove: ローカルストレージを使ったメモの機能を削除

- コンポーネントの分離に伴いメモ機能を`useMemo`に移行するため

kei615ykhm committed:
add: 簡易的なデザインを作成

kei615ykhm committed:
update: インポートのパスを変更

kei615ykhm committed:
add: #9 メモリストコンポーネント作成のためのひな形を生成

kei615ykhm committed:
update: 基本構造に整形

kei615ykhm committed:
add: `props`の定義

kei615ykhm committed:
add: 親コンポーネントから受け取ったメモの配置を表示させる記述

kei615ykhm committed:
wip

kei615ykhm committed:
add: `onDelete`プロパティの型定義

kei615ykhm committed:
docs: コメントでデータフローの説明を追加

kei615ykhm committed:
add: `MemoListProps`に`onDelete`プロパティを渡す

kei615ykhm committed:
add: `MemoList`コンポーネントを`page`に追加

kei615ykhm committed:
add: `MemoForm`コンポーネントとの間に余白を設定

- マージントップ

kei615ykhm committed:
add: メモとメモの間をグレーのラインで分けるスタイリング

kei615ykhm committed:
add: リスト内の余白を拡大し、カーソル時のアニメーションを追加

kei615ykhm committed:
add: 削除ボタンを赤くし、カーソル時に色を変化させるスタイル調整

kei615ykhm committed:
update: マージントップの幅を狭める調整

kei615ykhm committed:
docs: コンポーネント作成TODOコメントを削除

kei615ykhm committed:
add: メモフォームの追加ボタンをホバー時に青色に変化するように調整

kei615ykhm committed:
update: 視覚的にわかりやすく押しやすい追加ボタンに修正

- 青背景、白文字、ホバー時に濃い青色に変化させるように設定

kei615ykhm committed:
update: 視覚的にわかりやすく押しやすい削除ボタンに修正

- 赤背景、白文字、ホバー時に暗い赤色に変化させるように設定

kei615ykhm committed:
remove: インポートの誤植を削除

kei615ykhm committed:
remove: 機能していない`saveMemos`関数を削除

niaka3dayo commented on src/hooks/useMemos.ts:
```diff
@@ -0,0 +1,33 @@
+import { Memo } from "../types";
+
+// メモの状態を管理するカスタムフック
+export const useMemos = () => {
```
[MAY] Reactには元々`useMemo`という名前の組み込みHookがあるため、この命名は混乱を招くので、名前を変更したほうがいいです。

niaka3dayo commented on src/hooks/useMemos.ts:
```diff
@@ -0,0 +1,33 @@
+  }, []);
+
+  // メモを追加する関数
+  const addMemo = (content: string) => {
```
[May] ハンドラなので、`handleAddMemo`などの名前が一般的です。

niaka3dayo commented on src/hooks/useMemos.ts:
```diff
@@ -0,0 +1,33 @@
+  };
+
+  // 指定したIDのメモを削除する関数
+  const deleteMemo = (id: string) => {
```
同上

niaka3dayo commented on src/components/memos/MemoList.tsx:
```diff
@@ -0,0 +1,32 @@
+import React from "react";
+import { Memo } from "../../types";
+
+// MemoListPropsコンポーネントのプロパティの型定義
+interface MemoListProps {
+  memos: Memo[]; // メモの配列
+  onDelete: (id: string) => void; // メモ削除のための関数
```
[MAY] こういうコメントはTSDoc形式で書いておけば他の場所で使う際に、VSCodeではマウスオーバーでその説明を表示してくれるようになるはずなのでそのほうが便利です。

```suggestion
/** MemoListPropsコンポーネントのプロパティの型定義 */
interface MemoListProps {
  /** メモの配列 */
  memos: Memo[];
  /** メモ削除のための関数 */
  onDelete: (id: string) => void; 
```

niaka3dayo commented on src/components/memos/MemoForm.tsx:
```diff
@@ -0,0 +1,38 @@
+  // フォームの送信処理
+  const handleSubmit = (e: React.FormEvent) => {
+    e.preventDefault(); // デフォルトの送信処理をキャンセル
+    if (content.trim()) {
+      // 入力値が空のみである場合は何もせず処理を終了
+      onSubmit(content); // 親コンポーネントから渡されたonSubmit関数を実行
+      setContent(""); // 入力値をリセット
+    }
```
[SHOULD] `content.trim()`が返すのは文字列なので、この場合は`content`が空であったとしても空文字列を返すことになると思います。
確かに、空文字列を暗黙的にブール変換した場合は`false`になるのですが、暗黙的ブール変換はかなり可読性が悪く、不安なコードに見えます。
暗黙的ブール変換をやめ、早期リターンを使えばもう少し良いコードになると思います。

```suggestion
　　const trimedContent = content.trim();
　　
　　if(trimedContent.length < 1) {
　　  // 入力された文字が1文字もない場合は何もせず早期リターン
　　  return;
　　}
　　
　　onSubmit(content); // 親コンポーネントから渡されたonSubmit関数を実行
    setContent(""); // 入力値をリセット
```

参考:
https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/String/trim

kei615ykhm committed:
update: カスタムフックの命名を変更

- `useMemos`は誤解を招きやすいためReactが用意している`useMemo`に変更した

kei615ykhm committed:
update: メモを追加するハンドラの命名を修正

- ハンドラの命名は`handle`とつけるほうが一般的であるため

kei615ykhm committed:
update: メモを削除するハンドラの命名を修正

- ハンドラの命名は`handle`とつけるほうが一般的であるため

kei615ykhm committed:
rename: ファイル名を`useMemoManager`に変更

- カスタムフックの命名をビルトインフックと被らないように修正したためファイル名も併せて変更した

kei615ykhm committed:
update: ビルトインフックとの競合を防ぐために`useMemo`の命名を変更

- メモ一覧を管理するフックであるため`useMemoManager`と命名した

kei615ykhm committed:
docs: `MemoListProps`インターフェースと`onDelete`関数に`TSDoc`スタイルのコメントを採用

- `onDelete`関数には`@param`を使用して引数の説明を追加

kei615ykhm committed:
docs: `MemoList`コンポーネントに`TSDoc`スタイルのコメントを採用

- コンポーネントの役割と`props`の説明がマウスオーバーで表示されるようにした

kei615ykhm committed:
update: フォーム入力後の分岐処理を早期リターンさせる

- 暗黙的ブール変換では可読性が悪く、不安はコードに見えるため、条件が満たされない場合はすぐに処理を終了させる早期リターンが適切であるため

kei615ykhm committed:
update: クリーンな入力データを保持し、親コンポーネントに渡す役割を持つ`content`の命名を`trimmedContent`に統一

kei615ykhm committed:
docs: `MemoFormProps`インターフェースに`TSDoc`スタイルを採用し、`onSubmit`プロパティの説明を追加

kei615ykhm committed:
docs: `MemoForm`コンポーネントに`TSDoc`スタイルを採用し、その役割と`props`の説明を追加

kei615ykhm committed:
docs: `@remarks`を使用し、自己説明的なコメントの可読性を改善

kei615ykhm commented:
@niaka3dayo ご返信が遅れて申し訳ございません。前回のコードレビューでいただいたフィードバックに基づいて修正を行いました。変更点を一つのコメントにまとめましたので、ご確認ください。

主な変更点：

1. 命名の変更：
   - `useMemos` → `useMemoManager` (Reactの組み込みhookとの競合を避けるため)
   - `useMemos.ts` → `useMemoManager.ts`
   - `addMemo` → `handleAddMemo`
   - `deleteMemo` → `handleDeleteMemo`

2. src/components/memos/MemoForm.tsx:
   - 暗黙的ブール変換を早期リターンに変更
   - 注意：提案された`trimd`を`trimmed`に修正
   （ご提案いただいた命名を誤植と判断）

3. 全体的なコードリファクタリング：
   対象ファイル：MemoForm.tsx, MemoList.tsx, useMemoManager.ts
   - TSDoc形式のコメントを導入し、コンポーネントや関数の役割、パラメータ、返り値の説明を追加
   - 不要なコメントを削除し、コードの自己説明性を向上

4. useMemoManager.ts:
   - カスタムフック全体と各関数にTSDocコメントを追加
   - memosステートに簡単な説明を追加

特に注目していただきたい点：
- 各ファイルでのTSDocの導入
- MemoForm.tsxでの早期リターンの実装

全体的にご提案いただいた内容以上の修正を行っていますが、とくにTSDoc形式でコメントを残すアプローチは関数や`props`のデータフローを見失いやすい私にとっては大きな問題解決につながると感じて全体的なリファクタリングを行いました。

ご確認よろしくお願いいたします！

niaka3dayo commented on src/components/memos/MemoForm.tsx:
```diff
@@ -0,0 +1,62 @@
+/**
+ * メモを入力するためのフォームコンポーネント
+ * @param props - コンポーネントのプロパティ
+ * @param props.onSubmit - フォーム送信時に呼び出される関数
```
[MAY] `@param`や`@returns`は`TypeScript`を用いた`React Component`ではまず使わないです。
なぜなら、`Props`の型情報にTSDocを書いた段階でもう同じ情報だからです。

niaka3dayo commented on src/components/memos/MemoForm.tsx:
```diff
@@ -0,0 +1,62 @@
+ * @param props.onSubmit - フォーム送信時に呼び出される関数
+ */
+export const MemoForm: React.FC<MemoFormProps> = ({ onSubmit }) => {
+  /** メモの内容を管理するstate */
```
[MAY] `useState`で使う状態までコメントするのは、さすがにやりすぎです。見ればわかるようなものはコメントを追記するべきではなく、初見でわからないような複雑なものだけにコメントをつけるべきです。

`const [content, setContent] = React.useState<string>("");`

これは明らかに、`content`というステートを定義し、それは`string`型の状態で初期値が`""`であるというふうにコード上で明示しているので、この`TSDoc`は完全に蛇足です。

niaka3dayo commented on src/components/memos/MemoForm.tsx:
```diff
@@ -0,0 +1,62 @@
+ * @param props - コンポーネントのプロパティ
+ * @param props.onSubmit - フォーム送信時に呼び出される関数
+ */
+export const MemoForm: React.FC<MemoFormProps> = ({ onSubmit }) => {
```
[MAY] 最近はアロー関数を使ったこっちの書き方は主流ではないですね。パフォーマンスや、IDEでの使いやすさから、以下のような書き方を推奨します。

```suggestion
export default function MemoForm({ onSubmit }: MemoFormProps): React.ReactElement {
```

niaka3dayo commented on src/components/memos/MemoForm.tsx:
```diff
@@ -0,0 +1,62 @@
+       * @remarks
+       * - `value`: input要素の値をcontent状態と同期
+       * - `onChange`: 入力値が変更されたらcontent状態を更新
+       */}
```
[SHOULD] コンポーネントのリターンしている`children`の要素に対してのコメント、さすがにやりすぎです。
これをしたいなら、`input`を返すコンポーネントを切り分けて作ってください。

niaka3dayo commented on src/components/memos/MemoForm.tsx:
```diff
@@ -0,0 +1,62 @@
+       * メモ追加ボタン
+       * @remarks
+       * - `type="submit"`: フォームの送信をトリガー
+       */}
```
同上

niaka3dayo commented on src/components/memos/MemoList.tsx:
```diff
@@ -0,0 +1,40 @@
+ * @param props.memos - 表示するメモの配列
+ * @param props.onDelete - メモを削除するための関数
+ */
+export const MemoList: React.FC<MemoListProps> = ({ memos, onDelete }) => {
```
[MAY] 上で指摘した内容と同じです。

niaka3dayo commented on src/components/memos/MemoList.tsx:
```diff
@@ -0,0 +1,40 @@
+  /** メモの配列 */
+  memos: Memo[];
+  /**
+   * メモ削除のための関数
```
```suggestion
   * メモ削除のためのコールバック関数
```

niaka3dayo commented on src/components/memos/MemoForm.tsx:
```diff
@@ -0,0 +1,62 @@
+/** MemoFormコンポーネントのプロパティの型定義 */
+interface MemoFormProps {
+  /**
+   * フォーム送信時に呼び出される関数
```
```suggestion
   * フォーム送信時に呼び出されるコールバック関数
```

niaka3dayo commented on src/components/memos/MemoForm.tsx:
```diff
@@ -0,0 +1,62 @@
+   * フォームの送信を処理する関数
+   * @param e - フォーム送信イベント
+   */
+  const handleSubmit = (e: React.FormEvent) => {
```
[MAY] このままではこのコンポーネントが描画されるたびに、この関数も新しく描画され続けてしまうので、パフォーマンスの問題があります。`useCallback`でメモ化してください。

niaka3dayo commented on src/hooks/useMemoManager.ts:
```diff
@@ -0,0 +1,33 @@
+  }, []);
+
+  // メモを追加する関数
+  const handleAddMemo = (content: string) => {
+    const newMemo: Memo = {
+      id: Date.now().toString(),
+      content,
+      createdAt: new Date().toISOString(),
+    };
+    setMemos([...memos, newMemo]);
+  };
+
+  // 指定したIDのメモを削除する関数
+  const handleDeleteMemo = (id: string) => {
+    setMemos(memos.filter((memo) => memo.id !== id));
+  };
```
[MAY] ハンドラ各種、`useCallback`によるメモ化が必要です。また、なぜかこちらはTSDoc形式のコメントになっていないです。

niaka3dayo commented on src/hooks/useMemoManager.ts:
```diff
@@ -0,0 +1,33 @@
+import { useEffect, useState } from "react";
+import { Memo } from "../types";
+
+// メモの状態を管理するカスタムフック
```
[MAY] なぜかこちらは`TSDoc`形式になっていない

niaka3dayo commented on src/hooks/useMemoManager.ts:
```diff
@@ -0,0 +1,33 @@
+export const useMemoManager = () => {
+  const [memos, setMemos] = useState<Memo[]>([]);
+
+  // クライアントサイドでのみローカルストレージにメモを保存
```
[MAY] コメント内容が間違っています。
ここでは、`localStorage`に保存されたメモをロードしてるのであって、保存しているわけではないです。

niaka3dayo commented on src/hooks/useMemoManager.ts:
```diff
@@ -0,0 +1,33 @@
+  // クライアントサイドでのみローカルストレージにメモを保存
+  useEffect(() => {
+    const storedMemos = localStorage.getItem("memos");
+    if (storedMemos) {
```
[SHOULD] 暗黙的ブール変換をやめましょう。

```suggestion
    if (storedMemos !== undefined && storedMemos !== null && storedMemos !== '') {
```

niaka3dayo commented on src/hooks/useMemoManager.ts:
```diff
@@ -0,0 +1,33 @@
+  useEffect(() => {
+    const storedMemos = localStorage.getItem("memos");
+    if (storedMemos) {
+      setMemos(JSON.parse(storedMemos));
```
[SHOULD] 直接`JSON.parse`で`localStorage`内の文字列を展開して状態にセットしてますが、これだと想定外の値が入力されていた場合や、あとから仕様の変更があった際にオブジェクトを正しい形にアップデートしたいときなどに不具合が生じますね。

一度変数に`JSON.parse`で展開した後に、その変数の内容が想定と合っているかどうかを判定し、その後正しい形に修正してから状態にセットしたほうがいいです。不正な値であった場合のエラーハンドリングを意識しましょう。

niaka3dayo commented:
@kei615ykhm 見ました

kei615ykhm committed:
docs: 蛇足コメントを削除

kei615ykhm committed:
docs: メモ状態の管理に関するTSDocコメントを追加

kei615ykhm committed:
docs: ローカルストレージの初回レンダリング時に関するTSDocコメントを追加

kei615ykhm committed:
docs: メモを追加する関数に関するTSDocコメント

kei615ykhm committed:
docs: メモを削除する関数に関するTSDocコメント

kei615ykhm committed:
docs: 蛇足コメントを削除

kei615ykhm committed:
docs: メモ削除のコールバック関数を示すTSDocコメント

kei615ykhm committed:
docs: フォーム送信時に呼び出されるコールバック関数を示すTSDocコメント

kei615ykhm committed:
refactor: パフォーマンス向上のため、アロー関数ではなく関数宣言を使用してHomeコンポーネントを定義

kei615ykhm committed:
refactor: `export default Home;`を追加して、Homeコンポーネントをエクスポート

kei615ykhm committed:
refactor: インポート方法の変更

- `MemoForm.tsx`側で意図的な`children`プロップを定義するために型注釈を使用したエクスポート方法に変更した

kei615ykhm committed:
refactor: `handleSubmit`関数を`useCallback`でラップし、依存配列に`content`と`onSubmit`を追加

- この変更により、`content`と`onSubmit`が変更されない限り`handleSubmit`関数が再生成されなくなりパフォーマンスが改善することを期待する

kei615ykhm committed:
refactor: 値の有効性チェックにおける暗黙的なブール変換を明示的な条件式に変更

kei615ykhm committed:
add: 不正な値で会った場合のエラーハンドリングを追加

kei615ykhm committed:
docs: TODOコメントでzodを使ったMemoスキーマ定義に関する内容を記述

kei615ykhm committed:
docs: TODOコメントでパースとバリデーション追加に関する内容を記述

kei615ykhm committed:
docs: TODOコメントで新しいメモのバリデーションに関する内容を記述

kei615ykhm commented:
@niaka3dayo ご提案やご指摘をもとにコードを修正しました。次のステップとして、`zod`ライブラリを使ったバリデーションチェックを実装予定です。

ただし、現在のブランチでは多くの実装を行ってしまったため、一度`develop`にマージし、新しいブランチで #16 のタスクを進めようと思います。

現時点でのコードにはまだ課題が残っていますが、作業を次のフェーズに進めるために、私の判断でマージを行います。`zod`によるバリデーションチェックと型安全なコードへのリファクタリングが完了次第、再度コードレビューをお願いしたいと考えています。

これまで多くのサポートをいただき、ありがとうございました。
引き続き、よろしくお願い致します。
